src/transformStateWithClones.js

```diff
@@ -6,8 +6,49 @@
  *
  * @return {Object[]}
  */
+
+const TYPE_ADD_PROPERTIES = 'addProperties';
+const TYPE_REMOVE_PROPERTIES = 'removeProperties';
+const TYPE_CLEAR = 'clear';
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesList = [{ ...state }];
+
+  for (const action of actions) {
+    let currentObject = { ...statesList[statesList.length - 1] };
+
+    switch (action.type) {
+      case TYPE_ADD_PROPERTIES: {
+        const objectWithNewFields = action.extraData;
+
+        currentObject = {
+          ...currentObject,
+          ...objectWithNewFields,
+        };
+        break;
+      }
+
+      case TYPE_REMOVE_PROPERTIES: {
+        const keysToRemove = action.keysToRemove;
+
+        for (const key of keysToRemove) {
+          delete currentObject[key];
+        };
+        break;
+      }
+
+      case TYPE_CLEAR: {
+        currentObject = {};
+        break;
+      }
+
+      default: return 'error';
```

Comment:
Better will be to throw an error, not return string 
Like : "throw new Error(`Something went wrong: unexpected action.type: ${action.type}`);"